archinstall/lib/models/users.py

```diff
@@ -103,72 +107,150 @@ def _check_password_strength(
 		return PasswordStrength.VERY_WEAK
 
 
-_UserSerialization = TypedDict('_UserSerialization', {'username': str, '!password': str, 'sudo': bool})
+_UserSerialization = TypedDict(
+	'_UserSerialization',
+	{
+		'username': str,
+		'!password': NotRequired[str],
+		'sudo': bool,
+		'groups': list[str],
+		'enc_password': str | None
+	}
+)
 
 
-@dataclass
-class User:
-	username: str
-	password: str
-	sudo: bool
+class Password:
+	def __init__(
+		self,
+		plaintext: str = '',
+		enc_password: str | None = None
+	):
+		if plaintext:
+			enc_password = self._gen_yescrypt_hash(plaintext)
+
+		if not plaintext and not enc_password:
+			raise ValueError('Either plaintext or enc_password must be provided')
+
+		self._plaintext = plaintext
+		self.enc_password = enc_password
 
 	@property
-	def groups(self) -> list[str]:
-		# this property should be transferred into a class attr instead
-		# if it's every going to be used
-		return []
+	def plaintext(self) -> str:
+		return self._plaintext
 
-	def json(self) -> _UserSerialization:
-		return {
-			'username': self.username,
-			'!password': self.password,
-			'sudo': self.sudo
-		}
+	@plaintext.setter
+	def plaintext(self, value: str):
+		self._plaintext = value
+		self.enc_password = self._gen_yescrypt_hash(value)
 
-	@classmethod
-	def _parse(cls, config_users: list[_UserSerialization]) -> list['User']:
-		users = []
+	@override
+	def __eq__(self, other: object) -> bool:
+		if not isinstance(other, Password):
+			return NotImplemented
 
-		for entry in config_users:
-			username = entry.get('username', None)
-			password = entry.get('!password', '')
-			sudo = entry.get('sudo', False)
+		if self._plaintext and other._plaintext:
+			return self._plaintext == other._plaintext
 
-			if username is None:
-				continue
+		return self.enc_password == other.enc_password
 
-			user = User(username, password, sudo)
-			users.append(user)
+	def hidden(self) -> str:
+		if self._plaintext:
+			return '*' * len(self._plaintext)
+		else:
+			return '*' * 8
 
-		return users
+	@cached_property
+	def _libcrypt(self):
+		libc = ctypes.CDLL("libcrypt.so")
 
-	@classmethod
-	def _parse_backwards_compatible(cls, config_users: dict[str, dict[str, str]], sudo: bool) -> list['User']:
-		if len(config_users.keys()) > 0:
-			username = list(config_users.keys())[0]
-			password = config_users[username]['!password']
+		libc.crypt.argtypes = [ctypes.c_char_p, ctypes.c_char_p]
+		libc.crypt.restype = ctypes.c_char_p
+
+		libc.crypt_gensalt.argtypes = [ctypes.c_char_p, ctypes.c_int, ctypes.c_void_p, ctypes.c_size_t]
```

Comment:
The first argtype is correct but the rest are wrong. See the first quote from `crypt_gensalt(3)` in https://github.com/archlinux/archinstall/pull/3276#discussion_r2029082231 and https://github.com/archlinux/archinstall/pull/3276#discussion_r2029911717 for the correction.

Reply:
Arguments have been fixed